File: iOS/DataPool/Sources/DataPool/Infrastructure/PoolLayoutRemoteRepository.swift
import Core

public class PoolLayoutRemoteRepository : PoolLayoutRepository {
    private let poolLayoutDataSource: PoolLayoutRemoteDataSource

    private let networkErrorHandler: NetworkErrorHandler

    init(poolLayoutDataSource: PoolLayoutRemoteDataSource, networkErrorHandler: NetworkErrorHandler) {
        self.poolLayoutDataSource = poolLayoutDataSource
        self.networkErrorHandler = networkErrorHandler
    }

    public func getOpenPoolLayouts(next: String?, searchText: String?) async -> Result<Core.CursorPage<PoolLayout>, any Error> {
        return await networkErrorHandler.handle {
            let page = try await poolLayoutDataSource.getOpenPoolLayouts(next: next, searchText: searchText)
            return page.map { item in item.toPoolLayout() }
        }
    }
}

File: Android/data/pool/src/main/java/com/felipearpa/tyche/data/pool/infrastructure/PoolLayoutRemoteRepository.kt
package com.felipearpa.tyche.data.pool.infrastructure

import com.felipearpa.network.NetworkExceptionHandler
import com.felipearpa.tyche.core.paging.CursorPage
import com.felipearpa.tyche.core.paging.map
import com.felipearpa.tyche.data.pool.domain.PoolLayout
import com.felipearpa.tyche.data.pool.domain.PoolLayoutDataSource
import com.felipearpa.tyche.data.pool.domain.PoolLayoutRepository
import com.felipearpa.tyche.data.pool.domain.toPoolLayout

internal class PoolLayoutRemoteRepository(
    private val poolLayoutDataSource: PoolLayoutDataSource,
    private val networkExceptionHandler: NetworkExceptionHandler,
) : PoolLayoutRepository {
    override suspend fun getOpenPoolLayouts(
        next: String?,
        searchText: String?,
    ): Result<CursorPage<PoolLayout>> {
        return networkExceptionHandler.handle {
            poolLayoutDataSource.getOpenPoolLayouts(next = next, searchText = searchText)
                .map { poolLayoutResponse -> poolLayoutResponse.toPoolLayout() }
        }
    }
}
